include/picongpu/simulation/stage/AtomicPhysics.hpp

```diff
@@ -0,0 +1,604 @@
+            {
+                picongpu::particles::atomicPhysics::stage::DumpSuperCellDataToConsole<
+                    picongpu::particles::atomicPhysics::localHelperFields::LocalTimeStepField<picongpu::MappingDesc>,
+                    picongpu::particles::atomicPhysics::localHelperFields::PrintTimeStepToConsole>{}(
+                    mappingDesc,
+                    "LocalTimeStepField");
+            }
+
+        public:
+            AtomicPhysics() = default;
+
+            //! atomic physics stage sub-stage calls
+            void operator()(picongpu::MappingDesc const mappingDesc, uint32_t const currentStep) const
+            {
+                // debug only
+                using ForEachIonSpeciesDumpRateCacheToConsole = pmacc::meta::ForEach<
+                    AtomicPhysicsIonSpecies,
+                    particles::atomicPhysics::stage::DumpRateCacheToConsole<boost::mpl::_1>>;
+
+                pmacc::DataConnector& dc = pmacc::Environment<>::get().DataConnector();
+
+                // TimeRemainingSuperCellField
+                auto& localTimeRemainingField = *dc.get<S_TimeRemainingField>("LocalTimeRemainingField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutTimeRemaining
+                    = localTimeRemainingField.getGridLayout().sizeWithoutGuardND();
+
+                // FoundUnboundIonSuperCellField
+                auto& localFoundUnboundIonField = *dc.get<S_FoundUnboundField>("LocalFoundUnboundIonField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutFoundUnbound
+                    = localFoundUnboundIonField.getGridLayout().sizeWithoutGuardND();
+
+                // ElectronHistogramOverSubscribedSuperCellField
+                auto& localElectronHistogramOverSubscribedField
+                    = *dc.get<S_OverSubscribedField>("LocalElectronHistogramOverSubscribedField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutOverSubscription
+                    = localElectronHistogramOverSubscribedField.getGridLayout().sizeWithoutGuardND();
+
+                /// @todo find better way than hard code old value, Brian Marre, 2023
+                // `static` avoids that reduce is allocating each time step memory, which will reduce the performance.
+                static pmacc::device::Reduce deviceLocalReduce = pmacc::device::Reduce(static_cast<uint32_t>(1200u));
+
+                setTimeRemaining(); // = (Delta t)_PIC
+                // fix atomic state and charge state inconsistency
+                using ForEachIonSpeciesFixAtomicState = pmacc::meta::
+                    ForEach<AtomicPhysicsIonSpecies, particles::atomicPhysics::stage::FixAtomicState<boost::mpl::_1>>;
+                ForEachIonSpeciesFixAtomicState{}(mappingDesc);
+
+                uint16_t counterSubStep = 0u;
+                // atomicPhysics sub-stepping loop, ends when timeRemaining<=0._X
+                while(true)
```

Comment:
REFACTOR: I would have preferred an actual condition or even better a for-loop here. Reason:
 1. It is hard to comprehend, check and verify the control flow with all those hidden `break`
 statements scattered around.
 2. In my head the following three have different semantics:

    - while(true):
      Infinite event loop that's supposed to run forever or until terminated by an external event.
    - while(condition):
      An iterative algorithm that does not progress but has a boolean success criterion, e.g.,
      hit-or-miss algorithms that return on the first hit.
    - for(...)
      An iterative algorithm that progresses in some meaningful way (not necessarily by `++i`)

 Admittedly, the border between while(condition) and for(...) is relatively vague and there is room
 for interpretation and personal preference there. But while(true) clearly communicates something
 different.

Reply:
A for-loop is not possible, since we do not know in advance how many sub step we will do.

Fundamentally the entire loop body needs to be executed at least once therfore the best fit would be a do-while loop, but I am unsure how often people actually see do-whiles.

Open to changing it to do-while loops.

Reply:
refactor to use do-while loop

Reply:
Offline discussion yielded some good results but just for the record: A `for` loop is exactly the same as a `while` loop except for the fact that you have the option to give it a bit more structure by using the initialisation and increment field.

```C++
for(<initialisation>, <condition>, <increment>) {
    ...
}
```
is identical to
```C++
<initialisation>
while(<condition>) {
    ...
    <increment>
}
```
(except for scoping).

And the other way round
```C++
while(<condition>) {
    ...
}
```
is the same as
```C++
for (; <condition>; ) {
    ...
}
```

The choice is really about what you want to communicate.